### PR TITLE
Build conceptual and API docs with Sphinx

### LINKED ISSUE #28: Generate API Documentation

### PR DESCRIPTION
**Pull request by cgranade**
This PR adds support for building conceptual and API documentation using Sphinx, and for publishing the resulting documentation as a build artifact and to GitHub Pages. In particular, this PR focuses on setting up infrastructure and publishing pipelines for docs, but there's significant work remaining to make sure that API docs have complete coverage.

This PR is progress towards, but does not completely fix #28.

### CODE REVIEW

File: pyqir-jit/src/lib.rs [hunk 1]
```diff
@@ -4,6 +4,11 @@
 #![warn(clippy::all, clippy::pedantic)]
 // pyo3 generates errors with _obj and _tmp values
 #![allow(clippy::used_underscore_binding)]
+// NB: Some arguments get turned into Deref by PyO3, which is meaningless
+//     for Option of a type that's already Deref. We ignore the warning
+//     here since it's introduced by an upstream macro and not somethi8ng
```
[the remark this answers is not in the document]
Comment (cgranade) [reply pyqir-jit/src/lib.rs:9]: Ah, typos. Thanks for the catch!

File: docs/Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# Put it first so that "make" without argument is like "make help".
+help:
+	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+
+.PHONY: help Makefile
+
+# Catch-all target: route all unknown targets to Sphinx using the new
+# "make mode" option.  $(O) is meant as a shortcut for $(SPHINXOPTS).
+%: Makefile
+	@$(SPHINXBUILD) -M $@ "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
```
[the remark this answers is not in the document]
Comment (cgranade) [reply docs/Makefile:21]: These are created by Sphinx itself and are provided for consistency with other Sphinx projects. For the actual build, the existing psakefile.ps1 is used as per your suggestion.

File: eng/psakefile.ps1 [hunk 3]
```diff
@@ -330,3 +361,33 @@ task run-examples -Depends init {
...
+        $ArtifactPaths
+    )
+
+    Write-Host "##[info]Creating virtual environment for use with docs at $EnvironmentPath...";
+    python -m venv $EnvironmentPath
+
+    $activateScript = (Join-Path $EnvironmentPath "bin" "Activate.ps1");
+    if (-not (Test-Path $activateScript -ErrorAction SilentlyContinue)) {
+        Get-ChildItem $EnvironmentPath | Write-Host
+        throw "No activate script found for virtual environment at $EnvironmentPath; environment creation failed.";
```
[the remark this answers is not in the document]
Comment (cgranade) [reply eng/psakefile.ps1:380]:
```suggestion
        throw "No activate script found for virtual environment at $EnvironmentPath; environment creation failed."
```

File: docs/apiref/generator.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# pyqir_generator
```
[the remark this answers is not in the document]
Comment (cgranade) [reply docs/apiref/generator.md:1]: Makes sense, can go on and rename, then. More generally, this PR is focused on getting the docs infrastructure up, but I expect there to be a fair bit more docs authoring needed as the project continues.

File: eng/psakefile.ps1 [hunk 5]
```diff
@@ -330,3 +361,33 @@ task run-examples -Depends init {
...
+
+    & $activateScript;
+    try {
+        pip install -r $RequirementsPath;
+        foreach ($artifact in $ArtifactPaths) {
+            pip install $artifact
+        }
+    }
+    finally {
+        deactivate;
```
Comment (samarsha) [inline eng/psakefile.ps1:391]: By the way there are other semicolons on lines 90, 374, 377, 385, 391, not sure if you want to remove those too (sorry, I didn't mention them in my other comment). It's a minor issue though.

**Review by samarsha — APPROVED** [review]

File: eng/psakefile.ps1 (hunk 5 above)
Comment (cgranade) [reply eng/psakefile.ps1]: Thanks for the catch!

**Comment by cgranade** [comment]
@samarsha: Thanks for the review, Sarah! We should be good to merge soon, I think.

Looking into markdownlint and link checking in more detail, it seems two of the errors were actual errors caused by using ordinary links instead of Sphinx-aware links, while the rest were due to false positives (https://github.com/DavidAnson/markdownlint/issues/479 for upstream feature request). There may be a workaround for the false positives, will take a few minutes to try that out.

**Comment by cgranade** [comment]
@samarsha: To fix the CI warnings, I wound up having to fix the missing config files for markdownlint and cspell. I realize you approved, but the PR has changed a bit since then thanks to the CI issues; if you'd like to take another look, please let me know and I'll hold off on merging. Thanks!

**Review by samarsha — APPROVED** [review]

**Pull request merged**